src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2015.  Airbnb.com
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package com.airbnb.kafka;
+
+import com.airbnb.metrics.Dimension;
+import com.airbnb.metrics.NewStatsDReporter;
+import com.airbnb.metrics.StatsDReporter;
+
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import com.timgroup.statsd.NonBlockingStatsDClient;
+import com.timgroup.statsd.StatsDClient;
+import com.timgroup.statsd.StatsDClientException;
+import com.yammer.metrics.Metrics;
+import com.yammer.metrics.core.Gauge;
+import com.yammer.metrics.core.MetricName;
+import com.yammer.metrics.core.MetricsRegistry;
+import com.yammer.metrics.reporting.AbstractPollingReporter;
+import org.apache.kafka.common.metrics.KafkaMetric;
+import org.apache.kafka.common.metrics.MetricsReporter;
+import org.slf4j.LoggerFactory;
+
+public class NewStatsdMetricsReporter implements MetricsReporter {
+  private static final org.slf4j.Logger log = LoggerFactory.getLogger(StatsDReporter.class);
+
+  static final String STATSD_REPORTER_ENABLED = "external.kafka.statsd.reporter.enabled";
+  static final String STATSD_HOST = "external.kafka.statsd.host";
+  static final String STATSD_PORT = "external.kafka.statsd.port";
+  static final String STATSD_METRICS_PREFIX = "external.kafka.statsd.metrics.prefix";
+  static final String POLLING_INTERVAL_SECS = "kafka.metrics.polling.interval.secs";
+  static final String STATSD_DIMENSION_ENABLED = "external.kafka.statsd.dimension.enabled";
+
+  private boolean enabled;
+  private final AtomicBoolean running = new AtomicBoolean(false);
+  private String host;
+  private int port;
+  private String prefix;
+  private long pollingPeriodInSeconds;
+  private EnumSet<Dimension> metricDimensions;
+  private StatsDClient statsd;
+  private MetricsRegistry registry;
+  private Map<String, KafkaMetric> kafkaMetrics;
+
+  private AbstractPollingReporter underlying = null;
+
+  public boolean isRunning() {
+    return running.get();
+  }
+
+  @Override
+  public void init(List<KafkaMetric> metrics) {
```

Comment:
@parasitew this is not my understanding. 

- Based on the [documentation](https://github.com/apache/kafka/blob/0.9.0/clients/src/main/java/org/apache/kafka/common/metrics/MetricsReporter.java#L25), "This is called when the reporter is first registered to *initially register* all existing metrics". 
- This is confirmed by the `JmxReporter` implementation: https://github.com/apache/kafka/blob/0.9.0/clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java#L68-L69

The current implementation does not register anything on initialization, and any existing metrics will not be reported.

I think the confusion might come from the poor method name `metricChange`. The intent of that method is not to notify a change of the passed metric, but the creation of the passed metric.  `metricCreation` would have been more appropriate.

See also https://github.com/apache/kafka/blob/0.9.0/clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java#L249-L289